### PR TITLE
Avoid duplicate 3rd partires

### PR DESCRIPTION
**Pull request by wdammak**
https://github.com/Dolibarr/dolibarr/issues/19432

### CODE REVIEW

**Comment by hregis** [comment]
@wdammak #19432
What if the person has to create a company that has the same name but in a different city? How will it know that this customer is the right customer that has already been created?

**Comment by daraelmin** [comment]
Shouldn't we indicate a "method" to avoid dupplicates ?

Like an option in witch user can decide if it is a dupplicate if name and email are the same.

Or maybe an alert like "A comapny named **** has been already found in data base. Are you sure you want to create a second one named the same way ?"

**Comment by daraelmin** [comment]
We must keep in mind that third party can be individual and there can be a lot of john carpenter...

I would prefere a test based on email+name+firstname or an alert on email found (some familiy use the same email adress for all member just like postal adress)

And I ask my self what happend if the thirdparty is created by the member module ? Or public subscritpion link.

File: htdocs/societe/card.php [hunk 1]
```diff
@@ -1291,8 +1291,10 @@ function init_check_no_email(input) {
 			print '<span id="TypeName" class="fieldrequired">'.$form->editfieldkey('ThirdPartyName', 'name', '', $object, 0).'</span>';
 		}
 		print '</td><td'.(empty($conf->global->SOCIETE_USEPREFIX) ? ' colspan="3"' : '').'>';
-		print '<input type="text" class="minwidth300" maxlength="128" name="name" id="name" value="'.dol_escape_htmltag($object->name).'" autofocus="autofocus">';
-		print $form->widgetForTranslation("name", $object, $permissiontoadd, 'string', 'alpahnohtml', 'minwidth300');
+		//print '<input type="text" class="minwidth300" maxlength="128" name="name" id="name" value="'.dol_escape_htmltag($object->name).'" autofocus="autofocus">';
```
Comment (eldy) [inline htdocs/societe/card.php:1294]: I would prefer a more powerful implementation that will be an ajax check when we type the name of a company. After 3 chars entered, a javascript ajax call try to find some similar thirdparties and if found, show a warning popup "Warning: we found x thirdparties that looks similar: xxx, yyy, ..."
So we will be able to enhance the engine for "duplicate search" with more powerful rules (case insensitive, check even if typo, check ingoring accents, ...) than a simple combo with the rule "contains stubstring".

**Comment by josett225** [comment]
Hi There,

This is a real valuable feature requested to keep a database clean.
Just to add a comment/issue on ThirdParty duplicates, the contact import feature does not work with error messages due to this duplicates issue. After checking all the them are existing but are duplicated. Let me know if I need to open a ticket to clear the message and if the import can work on another key.
See below exemples:
 Ligne 9
    > Champ 32: 'COMPANY 1' n'est pas une référence existante comme Tiers
* Ligne 27
    > Champ 32: 'COMPANY 2' n'est pas une référence existante comme Tiers
* Ligne 28
    > Champ 32: 'COMPANY 2' n'est pas une référence existante comme Tiers
* Ligne 37
    > Champ 32: 'COMPANY 3' n'est pas une référence existante comme Tiers
* Ligne 39
    > Champ 32: 'COMPANY 4' n'est pas une référence existante comme Tiers
* Ligne 47
    > Champ 32: 'COMPANY 5' n'est pas une référence existante comme Tiers

**Comment by wdammak** [comment]
To please @eldy , here is a code draft for a powerful implementation:

> 
    $("#name").select2({
        ajax: {
          url: "htdocs/societe/card.php",
          dataType: 'json',
          delay: 250,
          data: function (params) {
                return {
                      q: params.term, // search term
                      page: params.page
                };
            },

            processResults: function (data, params) {
              // scrolling can be used
              params.page = params.page || 1;
              return {
                results: data.items,
                pagination: {
                  more: (params.page * 30) < data.total_count
                }
              };
            },
            cache: true
        },
        tags: true,
        placeholder: 'In the meantime we check if this 3rd already exists...',
        minimumInputLength: 1,
        templateResult: formatCustomer,
        templateSelection: formatCustomerSelection
    });

    function formatCustomer (Customer) {
        if (Customer.loading) {
            return Customer.text;    
        }

        var $container = $(
            "<div class='select2-result-repository clearfix'>" +
              "<div class='select2-result-repository__avatar'></div>" +
              "<div class='select2-result-repository__meta'>" +
                "<div class='select2-result-repository__title'></div>" +
                "<div class='select2-result-repository__ref'></div>" +
                "<div class='select2-result-repository__mail'>" +
                  "<div class='select2-result-repository__address'><i class='fa fa-flash'></i> </div>" +
                  "<div class='select2-result-repository__code_postal'><i class='fa fa-star'></i> </div>" +
                  "<div class='select2-result-repository__city'><i class='fa fa-star'></i> </div>" +
                  "<div class='select2-result-repository__country'><i class='fa fa-star'></i> </div>" +
                "</div>" +
              "</div>" +
            "</div>"
          );

        $container.find(".select2-result-repository__title").text(Customer.full_name);
        $container.find(".select2-result-repository__ref").text(Customer.ref);
        $container.find(".select2-result-repository__mail").append("Mail: " + Customer.mail);
        $container.find(".select2-result-repository__address").append("Address: " + Customer.address);
        $container.find(".select2-result-repository__code_postal").append("Code: " + Customer.code_postal);
        $container.find(".select2-result-repository__city").append("City: " + Customer.city);
        $container.find(".select2-result-repository__country").append("Country: " + Customer.country);

        return $container;
    }

    function formatCustomerSelection (Customer) {
        return Customer.full_name || Customer.text;
    }

**Comment by wdammak** [comment]
url: "htdocs/societe/card.php",
should be replaced with something like this:
url: "htdocs/core/ajax/thirdparty.php",

but maybe @eldy has other places already implemented to retrieve third party information in ajax

**Comment by eldy** [comment]
You can enhance 
url: "htdocs/core/ajax/ajaxcompanies.php"
to enhance it (currently it searches on a string in part of name, customer code or supplier code.

**Comment by wdammak** [comment]
> @eldy 
you mean url: "htdocs/scoiete/ajaxcompanies.php" ?
or do you want us to clone it in url: "htdocs/core/ajax/ajaxcompanies.php" and enhance it?

**Comment by wdammak** [comment]
the code draft for url: "htdocs/core/ajax/ajaxcompanies.php"

<?php
/* Copyright (C) 2006      Andre Cianfarani     <acianfa@free.fr>
 * Copyright (C) 2005-2009 Regis Houssin        <regis.houssin@inodbox.com>
 * Copyright (C) 2007-2010 Laurent Destailleur  <eldy@users.sourceforge.net>
 * Copyright (C) 2010      Cyrille de Lambert   <info@auguria.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

/**
 *       \file       htdocs/societe/ajaxcompanies.php
 *       \brief      File to return Ajax response on third parties request
 */

if (!defined('NOTOKENRENEWAL')) define('NOTOKENRENEWAL', 1); // Disables token renewal
if (!defined('NOREQUIREMENU'))  define('NOREQUIREMENU', '1');
if (!defined('NOREQUIREHTML'))  define('NOREQUIREHTML', '1');
if (!defined('NOREQUIREAJAX'))  define('NOREQUIREAJAX', '1');
if (!defined('NOREQUIRESOC'))   define('NOREQUIRESOC', '1');
if (!defined('NOCSRFCHECK'))    define('NOCSRFCHECK', '1');

require '../../main.inc.php';


/*
 * View
 */

// Ajout directives pour resoudre bug IE
//header('Cache-Control: Public, must-revalidate');
//header('Pragma: public');

//top_htmlhead("", "", 1);  // Replaced with top_httphead. An ajax page does not need html header.
top_httphead();

//print ''."\n";

dol_syslog(join(',', $_GET));


// Generation liste des societes
if (GETPOST('newcompany') || GETPOST('socid', 'int') || GETPOST('id_fourn'))
{
	$return_arr = array();

	// Define filter on text typed
	$socid = $_GET['newcompany'] ? $_GET['newcompany'] : '';
	if (!$socid) $socid = $_GET['socid'] ? $_GET['socid'] : '';
	if (!$socid) $socid = $_GET['id_fourn'] ? $_GET['id_fourn'] : '';

	$sql = "SELECT s.rowid, s.nom, s.name_alias, s.code_client, s.code_fournisseur, s.address, s.zip, s.town, s.email, s.siren, s.siret, s.ape, s.idprof4, s.client, s.fournisseur, s.datec";
	$sql .= " , p.label AS country, d.nom AS departement ";
	$sql .= " FROM ".MAIN_DB_PREFIX."societe as s";
	$sql .= " INNER JOIN ".MAIN_DB_PREFIX."c_country as p ON s.fk_pays = p.rowid";
	$sql .= " INNER JOIN ".MAIN_DB_PREFIX."c_departements as d ON s.fk_departement = d.code_departement";
	$sql .= " WHERE s.entity IN (".getEntity('societe').")";
	if ($socid)
	{
		$sql .= " AND (";
		// Add criteria on name/code
		if (!empty($conf->global->COMPANY_DONOTSEARCH_ANYWHERE))   // Can use index
		{
			$sql .= "nom LIKE '".$db->escape($socid)."%'";
			$sql .= " OR code_client LIKE '".$db->escape($socid)."%'";
			$sql .= " OR code_fournisseur LIKE '".$db->escape($socid)."%'";
		} else {
			$sql .= "nom LIKE '%".$db->escape($socid)."%'";
			$sql .= " OR code_client LIKE '%".$db->escape($socid)."%'";
			$sql .= " OR code_fournisseur LIKE '%".$db->escape($socid)."%'";
		}
		if (!empty($conf->global->SOCIETE_ALLOW_SEARCH_ON_ROWID)) $sql .= " OR rowid = '".$db->escape($socid)."'";
		$sql .= ")";
	}
	//if (GETPOST("filter")) $sql.= " AND (".GETPOST("filter", "alpha").")"; // Add other filters
	$sql .= " ORDER BY nom ASC";

	//dol_syslog("ajaxcompanies", LOG_DEBUG);
	$resql = $db->query($sql);
	if ($resql)
	{
		while ($row = $db->fetch_array($resql))
		{
			$label = $row['nom'];
			if ($socid) $label = preg_replace('/('.preg_quote($socid, '/').')/i', '<strong>$1</strong>', $label, 1);
			$row_array['label'] = $label;
			$row_array['value'] = $row['nom'];
			$row_array['key'] = $row['rowid'];
			$row_array['name_alias'] = $row['name_alias'];
			$row_array['code_client'] = $row['code_client'];
			$row_array['code_fournisseur'] = $row['code_fournisseur'];
			$row_array['address'] = $row['address'];
			$row_array['zip'] = $row['zip'];
			$row_array['town'] = $row['town'];
			$row_array['email'] = $row['email'];
			$row_array['siren'] = $row['siren'];
			$row_array['siret'] = $row['siret'];
			$row_array['ape'] = $row['ape'];
			$row_array['idprof4'] = $row['idprof4'];
			$row_array['client'] = $row['client'];
			$row_array['fournisseur'] = $row['fournisseur'];
			$row_array['datec'] = $row['datec'];
			$row_array['country'] = $row['country'];
			$row_array['departement'] = $row['departement'];

			array_push($return_arr, $row_array);
		}

		echo json_encode($return_arr);
	} else {
		echo json_encode(array('nom'=>'Error', 'label'=>'Error', 'key'=>'Error', 'value'=>'Error'));
	}
} else {
	echo json_encode(array('nom'=>'ErrorBadParameter', 'label'=>'ErrorBadParameter', 'key'=>'ErrorBadParameter', 'value'=>'ErrorBadParameter'));
}

**Comment by wdammak** [comment]
New code (tested is OK But it needs to be improved and the display taken care of for it to be very pretty)
image

**I am counting on the dolibarr experts to finalize and improve it because I am abroad...**

file societe/card.php

		print '</td><td'.(empty($conf->global->SOCIETE_USEPREFIX) ? ' colspan="3"' : '').'>';
		//print '<input type="text" class="minwidth300" maxlength="128" name="name" id="name" value="'.dol_escape_htmltag($object->name).'" autofocus="autofocus">';
		//print $form->widgetForTranslation("name", $object, $permissiontoadd, 'string', 'alpahnohtml', 'minwidth300');
		// print $form->select_company(0, 'name', '(s.client = 1 OR s.client = 3) AND status=1', ' ');
		print '<select class="name" name="name" id="name" style="min-width:400px"></select>';
		?>
		<script>	
		$(document).ready(function () { 
			$('#name').select2({
				ajax: {
				  url: '/htdocs/core/ajax/ajaxcompanies.php',
				  dataType: 'json',
				  delay: 250,
				  data: function (params) {
						return {
							newcompany: params.term // search term
						};
				  },
				  processResults: function (data, params) {
					  return {
						results: data
					  };
				  },
				  cache: true
				},
				
				placeholder: 'In the meantime we check if this 3rd already exists...',
				minimumInputLength: 3,
				width: 'resolve',		/* off or resolve */
				language: select2arrayoflanguage,
				containerCssClass: ':all:',	
				selectionCssClass: ':all:',
				tags: true,
				multiple: true,
				templateResult: formatCustomer,
				templateSelection: formatCustomerSelection
			});

			function formatCustomer (Customer) {
				
				if (Customer.loading) {
					return Customer.text;
				}

				var $container = $("<div class='select2-result-repository clearfix'>" +
					 "<div class='select2-result-repository__avatar'></div>" +
					  "<div class='select2-result-repository__meta'>" +
						"<div class='select2-result-repository__title'></div>" +
						"<div class='select2-result-repository__name_alias'></div>" +
						"<div class='select2-result-repository__code_client'></div>" +
						"<div class='select2-result-repository__code_fournisseur'></div>" +
						"<div class='select2-result-repository__companies_info'>" +
						  "<div class='select2-result-repository__email'><i class='fa fa-at'></i> </div>" +
						  "<div class='select2-result-repository__address'><i class='fa fa-home'></i> </div>" +
						  "<div class='select2-result-repository__country'><i class='fa fa-flag'></i> </div>" +
						  "<div class='select2-result-repository__departement'><i class='fa fa-circle-o'></i> </div>" +
						  "<div class='select2-result-repository__zip'><i class='fa fa-circle-o'></i> </div>" +
						  "<div class='select2-result-repository__town'><i class='fa fa-circle-o'></i> </div>" +
						  "<div class='select2-result-repository__siren'><i class='fa fa-circle-o'></i> </div>" +
						  "<div class='select2-result-repository__datec'><i class='fa fa-calendar'></i> </div>" +
						"</div>" +
					  "</div>" +
					"</div>"
				  );

				$container.find('.select2-result-repository__title').text(Customer.label);
				$container.find('.select2-result-repository__name_alias').text(Customer.name_alias);
				$container.find('.select2-result-repository__code_client').text(Customer.code_client);
				$container.find('.select2-result-repository__code_fournisseur').text(Customer.code_fournisseur);
				$container.find('.select2-result-repository__email').append('Mail: ' + Customer.email);
				$container.find('.select2-result-repository__address').append('Address: ' + Customer.address);
				$container.find('.select2-result-repository__country').append('Country: ' + Customer.country);
				$container.find('.select2-result-repository__departement').append('Departement: ' + Customer.departement);
				$container.find('.select2-result-repository__zip').append('Zip: ' + Customer.zip);
				$container.find('.select2-result-repository__town').append('Town: ' + Customer.town);
				$container.find('.select2-result-repository__siren').append('Siren: ' + Customer.siren);
				$container.find('.select2-result-repository__datec').append('Created: ' + Customer.datec);

				return $container;
			}

			function formatCustomerSelection (selection) {
				return selection.label || selection.text;
			}
		});
		</script>


Add the file "htdocs/core/ajax/ajaxcompanies.php" with the content



	/**
	 *       \file       htdocs/core/ajax/ajaxcompanies.php
	 *       \brief      File to return Ajax response on third parties request
	 */

	if (!defined('NOTOKENRENEWAL')) define('NOTOKENRENEWAL', 1); // Disables token renewal
	if (!defined('NOREQUIREMENU'))  define('NOREQUIREMENU', '1');
	if (!defined('NOREQUIREHTML'))  define('NOREQUIREHTML', '1');
	if (!defined('NOREQUIREAJAX'))  define('NOREQUIREAJAX', '1');
	if (!defined('NOREQUIRESOC'))   define('NOREQUIRESOC', '1');
	if (!defined('NOCSRFCHECK'))    define('NOCSRFCHECK', '1');

	require '../../main.inc.php';


	/*
	 * View
	 */

	// Ajout directives pour resoudre bug IE
	//header('Cache-Control: Public, must-revalidate');
	//header('Pragma: public');

	//top_htmlhead("", "", 1);  // Replaced with top_httphead. An ajax page does not need html header.
	top_httphead();

	//print ''."\n";

	dol_syslog(join(',', $_GET));


	// Generation liste des societes
	if (GETPOST('newcompany') || GETPOST('socid', 'int') || GETPOST('id_fourn'))
	{
		$return_arr = array();

		// Define filter on text typed
		$socid = $_GET['newcompany'] ? $_GET['newcompany'] : '';
		if (!$socid) $socid = $_GET['socid'] ? $_GET['socid'] : '';
		if (!$socid) $socid = $_GET['id_fourn'] ? $_GET['id_fourn'] : '';

		$sql = "SELECT s.rowid, s.nom, s.name_alias, s.code_client, s.code_fournisseur, s.address, s.zip, s.town, s.email, s.siren, s.siret, s.ape, s.idprof4, s.client, s.fournisseur, s.datec";
		$sql .= " , c.label as country, d.nom as departement";
		$sql .= " FROM ".MAIN_DB_PREFIX."societe as s";
		$sql .= " LEFT JOIN ".MAIN_DB_PREFIX."c_country as c ON c.rowid = s.fk_pays";
		$sql .= " LEFT JOIN ".MAIN_DB_PREFIX."c_departements as d ON d.rowid = s.fk_departement";
		$sql .= " WHERE s.entity IN (".getEntity('societe').")";
		if ($socid)
		{
			$sql .= " AND (";
			// Add criteria on name/code
			if (!empty($conf->global->COMPANY_DONOTSEARCH_ANYWHERE))   // Can use index
			{
				$sql .= "s.nom LIKE '".$db->escape($socid)."%'";
				$sql .= " OR s.code_client LIKE '".$db->escape($socid)."%'";
				$sql .= " OR s.code_fournisseur LIKE '".$db->escape($socid)."%'";
			} else {
				$sql .= "s.nom LIKE '%".$db->escape($socid)."%'";
				$sql .= " OR s.code_client LIKE '%".$db->escape($socid)."%'";
				$sql .= " OR s.code_fournisseur LIKE '%".$db->escape($socid)."%'";
			}
			if (!empty($conf->global->SOCIETE_ALLOW_SEARCH_ON_ROWID)) $sql .= " OR s.rowid = '".$db->escape($socid)."'";
			$sql .= ")";
		}
		//if (GETPOST("filter")) $sql.= " AND (".GETPOST("filter", "alpha").")"; // Add other filters
		$sql .= " ORDER BY s.nom ASC";

		//dol_syslog("ajaxcompanies", LOG_DEBUG);
		$resql = $db->query($sql);
		
		if ($resql)
		{
			while ($row = $db->fetch_array($resql))
			{
				$label = $row['nom'];
				if ($socid) $label = $label;
				$row_array['label'] = $label;
				$row_array['value'] = $row['nom'];
				$row_array['key'] = $row['rowid'];
				$row_array['name_alias'] = $row['name_alias'];
				$row_array['code_client'] = $row['code_client'];
				$row_array['code_fournisseur'] = $row['code_fournisseur'];
				$row_array['address'] = $row['address'];
				$row_array['zip'] = $row['zip'];
				$row_array['town'] = $row['town'];
				$row_array['email'] = $row['email'];
				$row_array['siren'] = $row['siren'];
				$row_array['siret'] = $row['siret'];
				$row_array['ape'] = $row['ape'];
				$row_array['idprof4'] = $row['idprof4'];
				$row_array['client'] = $row['client'];
				$row_array['fournisseur'] = $row['fournisseur'];
				$row_array['datec'] = $row['datec'];
				$row_array['country'] = $row['country'];
				$row_array['departement'] = $row['departement'];

				array_push($return_arr, $row_array);
			}

			echo json_encode($return_arr);
		} else {
			echo json_encode(array('nom'=>'Error', 'label'=>'Error', 'key'=>'Error', 'value'=>'Error'));
		}
	} else {
		echo json_encode(array('nom'=>'ErrorBadParameter', 'label'=>'ErrorBadParameter', 'key'=>'ErrorBadParameter', 'value'=>'ErrorBadParameter'));
	}

**Comment by wdammak** [comment]
rendering of the latest changes
image

**Pull request merged**